Currency calculation logic modified, added comments

Amarok121 opened the pull request:
설명 : Modified currency calculation logic

변경 파일1: RoundState.java

int coin = 0;
        int baseCoin = roundScore / 10;
        //레벨 보너스(levelBonus)
        int levelBonus = baseCoin * currState.getLevel();
        coin += levelBonus;

New variables for containing other bonuses

Amarok121 committed:
Coin logic modified

Amarok121 committed:
Coin logic modification with extra comments
Author: Kwon Ik Ju
Date: 2024.10.10
Update: More precise comments for the modified method + new variables

Edited File: RoundState.java

suhynnoh commented on src/CtrlS/RoundState.java:
```diff
@@ -35,23 +35,37 @@ public RoundState(GameState prevState, GameState currState) {
+        //정확도 보너스(accuracyBonus)
+        int accuracyBonus = 0;
 
         if (roundHitRate > 0.9) {
+            accuracyBonus += (int) (coin * 0.3);
             coin += (int) (coin * 0.3); // 30% 보너스 지급
             Core.getLogger().info("hitRate bonus occurs (30%).");
         } else if (roundHitRate > 0.8) {
+            accuracyBonus += (int) (coin * 0.3);
             coin += (int) (coin * 0.2); // 20% 보너스 지급
             Core.getLogger().info("hitRate bonus occurs (20%).");
+        } else if (roundHitRate > 0.7) {
+            accuracyBonus += (int) (coin * 0.3);
+            coin += (int) (coin * 0.1); // 10% 보너스 지급
+            Core.getLogger().info("hitRate bonus occurs (10%).");
         }
```
현재 코드는 레벨 보너스가 포함된 coin에서 퍼센트를 계산하는 문제가 아직 있는 것 같습니다.
그리고 accuracyBonus는 계산만 되고 사용이 되지 않는 것 같습니다.

아래와 같이 accuracyBonus를 baseCoin을 기반으로 계산한 뒤, 조건문 밖에서 coin에 accuracyBonus를 더하는 게 어떨까요?
```suggestion
        if (roundHitRate > 0.9) {
            accuracyBonus += (int) (baseCoin * 0.3);
            Core.getLogger().info("hitRate bonus occurs (30%).");
        } else if (roundHitRate > 0.8) {
            accuracyBonus += (int) (baseCoin * 0.2);
            Core.getLogger().info("hitRate bonus occurs (20%).");
        } else if (roundHitRate > 0.7) {
            accuracyBonus += (int) (baseCoin * 0.1);
            Core.getLogger().info("hitRate bonus occurs (10%).");
        }
        coin += accuracyBonus
```

Koominjae replied:
> 현재 코드는 레벨 보너스가 포함된 coin에서 퍼센트를 계산하는 문제가 아직 있는 것 같습니다. 그리고 accuracyBonus는 계산만 되고 사용이 되지 않는 것 같습니다.
> 
> 아래와 같이 accuracyBonus를 baseCoin을 기반으로 계산한 뒤, 조건문 밖에서 coin에 accuracyBonus를 더하는 게 어떨까요?

동의 합니다. 한 보너스가 다른 보너스 계산에 개입이 되서는 안되고 보너스 계산은 오로지 baseCoin으로만 이루어져 따로 변수에 저장되고 baseCoin에 각 보너스가 더해져 최종 해당 라운드에서의 획득 코인이 되어야 합니다.

Amarok121 replied:
확인했습니다. 제안 사항을 커밋에 반영했습니다.

Koominjae commented:
추가로 클리어 하지 못하고 게임 오버시 해당 라운드에서 적 함선을 처치해 얻은 코인을 제외한 `timeBonus`, `levelBonus`, `accuracyBonus`는 지급하지 않기로 했었는데 이 로직은 어떻게 구현하실 예정이신가요?

Amarok121 committed:
Update src/CtrlS/RoundState.java

Co-authored-by: suhynnoh <suhynnoh@hanyang.ac.kr>

Amarok121 committed:
Update RoundState.java

Geguuuul commented on src/CtrlS/RoundState.java:
84라인의 levelBonus2는 기존의 levelBonus와 다른건가요?

Geguuuul replied:
아 이해했어요

Amarok121 committed:
feat: RoundState modification for checking life
- Added a new variable 'roundHealth'
- It is for containing current gamestate's remaining life
- If it's below 1, level/acc/time bonus will be skipped

Amarok121 commented:
> 추가로 클리어 하지 못하고 게임 오버시 해당 라운드에서 적 함선을 처치해 얻은 코인을 제외한 `timeBonus`, `levelBonus`, `accuracyBonus`는 지급하지 않기로 했었는데 이 로직은 어떻게 구현하실 예정이신가요?

우선 현재 체력을 저장하기 위한 roundHealth 변수를 따로 만들어서, 게임 오버 시(체력 1 미만) 해당 보너스들을 패스하도록 조건문을 작성해 두었습니다.

suhynnoh commented on src/CtrlS/RoundState.java:
```diff
@@ -67,20 +82,31 @@ private int calculateCoin() {
         }
 
+        //코인 증가
+        if(roundHealth > 0){
```
roundHealth 변수와 getter를 만들지 않고, 단순화 하는 게 어떨까요?

```suggestion
        if(currState.getLivesRemaining() > 0){
```

Amarok121 committed:
Update src/CtrlS/RoundState.java

Simplification accepted.

Co-authored-by: suhynnoh <suhynnoh@hanyang.ac.kr>

suhynnoh commented on src/CtrlS/RoundState.java:
```diff
@@ -9,6 +9,8 @@ public class RoundState {
     private final GameState prevState;
     private final GameState currState;
     private final int roundScore;
+
+    private final int roundHealth;
```
```suggestion

```

suhynnoh commented on src/CtrlS/RoundState.java:
```diff
@@ -23,6 +25,7 @@ public RoundState(GameState prevState, GameState currState) {
         this.prevState = prevState;
         this.currState = currState;
         this.roundScore = currState.getScore() - prevState.getScore();
+        this.roundHealth = currState.getLivesRemaining();
```
```suggestion
```

suhynnoh commented on src/CtrlS/RoundState.java:
```diff
@@ -67,20 +82,31 @@ private int calculateCoin() {
         return roundScore;
     }
 
+    public int getRoundHealth(){
+        return roundHealth;
+    }
+
```
```suggestion
```

Amarok121 committed:
Update src/CtrlS/RoundState.java

Co-authored-by: suhynnoh <suhynnoh@hanyang.ac.kr>

Amarok121 committed:
Update src/CtrlS/RoundState.java

Co-authored-by: suhynnoh <suhynnoh@hanyang.ac.kr>

Amarok121 committed:
Update src/CtrlS/RoundState.java

Co-authored-by: suhynnoh <suhynnoh@hanyang.ac.kr>
